### PR TITLE
missed adding electra process epoch on state replay - devnet 1

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

The core logic of process slots was duplicated in the replay function,so we forgot to update epoch processing in the replay function for the electra process slot. This PR refactors the core logic to be reusable and adds the corrected electra process epoch

**Which issues(s) does this PR fix?**

Fixes #

**Other notes for review**

### CODE REVIEW

File: beacon-chain/core/transition/transition.go [hunk 1]
```diff
@@ -223,50 +212,68 @@ func ProcessSlots(ctx context.Context, state state.BeaconState, slot primitives.
...
 		}
 	} else if err != nil {
 		return nil, err
 	}
 	defer func() {
 		SkipSlotCache.MarkNotInProgress(key)
 	}()
-
-	for state.Slot() < slot {
+	customErrFn := func(ctx context.Context, st state.BeaconState) error {
```
Comment (james-prysm) [inline beacon-chain/core/transition/transition.go:223]: reviewer please carefully look at the use of this function and let me know if this works well for our usecase

File: beacon-chain/core/transition/transition.go [hunk 2]
```diff
@@ -173,35 +173,24 @@ func ProcessSlotsIfPossible(ctx context.Context, state state.BeaconState, target
...
-//	  assert state.slot < slot
-//	  while state.slot < slot:
-//	      process_slot(state)
-//	      # Process epoch on the start slot of the next epoch
-//	      if (state.slot + 1) % SLOTS_PER_EPOCH == 0:
-//	          process_epoch(state)
-//	      state.slot = Slot(state.slot + 1)
-func ProcessSlots(ctx context.Context, state state.BeaconState, slot primitives.Slot) (state.BeaconState, error) {
+// ProcessSlots includes core slot processing as well as a cache
+func ProcessSlots(ctx context.Context, beaconState state.BeaconState, slot primitives.Slot) (state.BeaconState, error) {
```
Comment (potuz) [inline beacon-chain/core/transition/transition.go:177]: Why this change?

File: beacon-chain/core/transition/transition.go [hunk 3]
```diff
@@ -231,42 +222,63 @@ func ProcessSlots(ctx context.Context, state state.BeaconState, slot primitives.
...
-					return nil, errors.Wrap(err, fmt.Sprintf("could not process %s epoch", version.String(state.Version())))
-				}
-			} else {
-				if err = electra.ProcessEpoch(ctx, state); err != nil {
-					tracing.AnnotateError(span, err)
-					return nil, errors.Wrap(err, fmt.Sprintf("could not process %s epoch", version.String(state.Version())))
-				}
-			}
+
+		state, err = ProcessEpoch(ctx, state)
```
Comment (potuz) [inline beacon-chain/core/transition/transition.go:276]: What happens for Phase0?

File: beacon-chain/core/transition/transition.go [hunk 4]
```diff
@@ -278,12 +290,29 @@ func ProcessSlots(ctx context.Context, state state.BeaconState, slot primitives.
...
+	return state, nil
+}
 
-	if highestSlot < state.Slot() {
-		SkipSlotCache.Put(ctx, key, state)
+// ProcessEpoch is a wrapper on fork specific epoch processing
+func ProcessEpoch(ctx context.Context, state state.BeaconState) (state.BeaconState, error) {
+	var err error
+	if time.CanProcessEpoch(state) {
+		if state.Version() == version.Phase0 {
```
Comment (potuz) [inline beacon-chain/core/transition/transition.go:300]: can we switch the order of these ones?

File: beacon-chain/core/transition/transition.go (hunk 3 above)
Comment (potuz) [reply beacon-chain/core/transition/transition.go:276]: discussed offline, this is fine

File: beacon-chain/core/transition/transition.go [hunk 5]
```diff
@@ -231,42 +222,63 @@ func ProcessSlots(ctx context.Context, state state.BeaconState, slot primitives.
...
+//	      process_slot(state)
+//	      # Process epoch on the start slot of the next epoch
+//	      if (state.slot + 1) % SLOTS_PER_EPOCH == 0:
+//	          process_epoch(state)
+//	      state.slot = Slot(state.slot + 1)
+func ProcessSlotsCore(ctx context.Context, span *trace.Span, state state.BeaconState, slot primitives.Slot, fn customProcessingFn) (state.BeaconState, error) {
+	var err error
+	for state.Slot() < slot {
+		if fn != nil {
+			if err = fn(ctx, state); err != nil {
```
Comment (potuz) [inline beacon-chain/core/transition/transition.go:265]:
```suggestion
if fn != nil && (err = fn(ctx, state)) != nil {
```

File: beacon-chain/core/transition/transition.go [hunk 6]
```diff
@@ -231,42 +222,63 @@ func ProcessSlots(ctx context.Context, state state.BeaconState, slot primitives.
 	defer func() {
 		SkipSlotCache.MarkNotInProgress(key)
 	}()
+	state, err = ProcessSlotsCore(ctx, span, state, slot, cacheBestBeaconStateOnErrFn(highestSlot, key))
```
Comment (james-prysm) [inline beacon-chain/core/transition/transition.go:225]: please look at the use of cacheBestBeaconStateOnErrFn carefully

**Review by potuz — APPROVED** [review]
